Adds TypeScript module declaration

ccampanale opened the pull request:
- this was manually generated using the API documentation on the Atlas
documentation site as referenced, per each request, in this projects
readme

ccampanale commented:
Hi @montumodi, please review this PR and included TypeScript declarations and let me know if you have any questions. I'm not entirely sure it's perfect as I created this manually but it's been working for my uses thus far. I originally created this as a type definition stored locally in an internal package/library for my organization but am running into complexities with dependent packages being able to resolve these dependencies and compile successfully. I figured I would attempt to push a PR here first before adding to https://github.com/DefinitelyTyped/DefinitelyTyped.

montumodi commented:
Thanks @ccampanale Really appreciate this. I am not that good with typescript. I have some questions

I am trying to use it like this

```
import getMongodbAtlasApiClient, {AtlasClient} from "mongodb-atlas-api-client";
const client: AtlasClient = getMongodbAtlasApiClient({
  publicKey: "",
  privateKey: "",
  baseUrl: "",
  projectId: ""
});
```

I am getting error `TypeError: mongodb_atlas_api_client_1.default is not a function`. Do you have any idea why it is so? Do I need to add some setting in tsconfig while compiling this?

montumodi commented:
@ccampanale I guess we need export function as `default` as well to make it work. basically in index.js we need to add another line.

```
module.exports.default = getMongodbAtlasApiClient;
```

montumodi commented on index.d.ts:
```diff
@@ -0,0 +1,573 @@
+declare module 'mongodb-atlas-api-client' {
+
```
Should we add this `index.d.ts` in the src folder? Also, will it better to have separate type file for each entity. Like `user.d.ts`, `alert.d.ts` etc. It will be more managable.  And then in `index.d.ts` refer these files. What are your thoughts on this?

ccampanale replied:
I'm only just starting to get my hands dirty with TS declaration files so I"m not 100% sure. From what I've seen so far, these are usually included at the root of the module, but there are also times where there are declarations throughout the source as well. I don't think there would be an issue with splitting into different files but I think there might need to be some changes to the declaration, maybe as a namespace across multiple files instead of a module?

Still playing locally to see if I can answer you questions above.

montumodi commented on index.d.ts:
```diff
@@ -0,0 +1,573 @@
+
+    // AlertClient
+    export type AlertId = string;
+    export interface AlertClient {
```
should this be just called `Alert` instead of `AlertClient` (and other entities as well) as it denotes an entity?

ccampanale replied:
I think I see where you are coming from, if each property of the `AtlasClient` is essentially an entity endpoint, we could drop the `*Client` nomenclature and just call them what they are. However, for consistency, if you want to change the `AlertClient` I think it would make sense to adjust them all.

ccampanale commented:
I think the issues we are seeing stem from the fact that I created the declaration file as a "module declaration" in a special `./src/types` directory in a project within my organization before moving it to the root of your project and pushing the PR. I'm still getting used to some of this with TypeScript so some things go over my head but I think we can get to the bottom of it.

I'm going to try and re-work some things locally and if I can get things working I will push up some changes.

ccampanale commented:
Just pushed a commit that I believe cleans things up a bit. I renamed the entity interfaces that make up the `AtlasClient` as you suggested and split this into separate files. I think there is a way to pull these all into a single namespace to avoid all the imports across files but I'm only just getting started with this side of TypeScript. 🤷‍♂ 

Testing locally, I can install this package as-is into one of my packages (symlinked) and my existing TypeScript code compiles just fine.

> @ccampanale I guess we need export function as `default` as well to make it work. basically in index.js we need to add another line.
> 
> ```
> module.exports.default = getMongodbAtlasApiClient;
> ```

Regarding this, the declaration file exports the factory function `getMongodbAtlasApiClient()` as the default function because that is how it is exported on the JavaScript side:

```
module.exports = getMongodbAtlasApiClient;
```

Don't think anything needs to change in that regard.

Not sure if/how you want to include tests for this - or even rewrite your current tests in TS - but I will help however I can. So far just some local testing seems to be working fine.

montumodi commented on src/index.d.ts:
```diff
@@ -0,0 +1,78 @@
+import {Alert} from './alert'
```
should these imports be `export * from './alert'` so that it exports everything from alert typing. At the moment import line doesn't do anything. I guess once we fix this, rest looks good.

ccampanale replied:
I believe there is a better approach with this that puts all of the types into the same namespace across multiple files but I'm not entirely sure and I'm still trying to understand how that works.

When I split these and tested locally it worked fine but now TypeScript is complaining because, as you pointed out, several types are not re-exported in the index and are unavailable.

They are imported here in order to provide the types to the main `AtlasClient` that collects them but, as you pointed out, there is an issue. Not sure why TS decided to not complain last night...

montumodi replied:
ya. I faced the same issue. When I tried with `export *` it all worked. I found this solution on some of the links I googled.

ccampanale replied:
Yeah, most of my experience with this is in packages where all the types are in a single declaration file in the root, especially for packages which are not written in TS but have a declaration file for dependents which are written in TS. TS packages which I've written usually co-locate the types and interfaces throughout the codebase so writing declarations is still foreign to me.

ccampanale commented:
The last commit re-exports the declarations and the following appears to compile fine:
```typescript
import getMongodbAtlasApiClient, { AtlasClient, CreateUserRequest, AtlasError, GetUserResponse } from '..';

const client: AtlasClient = getMongodbAtlasApiClient({
    privateKey: '',
    publicKey: '',
    projectId: '',
    baseUrl: ''
});

const newUser: CreateUserRequest = {
    databaseName: 'db',
    groupId: 'myGroup',
    roles: [{
        databaseName: 'db',
        roleName: 'readWrite'
    }],
    password: 'usersPassword',
    username: 'aUser'
}

client.user.create(newUser).then(response => {
    if ((response as AtlasError).errorCode) {
        throw new Error('failed to create user');
    }
    const createdUser = (response as GetUserResponse);
    console.log('created new user with access to: ', createdUser.databaseName)
});
```

(My primary projects is still complaining but I think TypeScript is just confused in that case... some of this is still very confusing to me. haha)

montumodi commented:
> The last commit re-exports the declarations and the following appears to compile fine:
> 
> ```ts
> import getMongodbAtlasApiClient, { AtlasClient, CreateUserRequest, AtlasError, GetUserResponse } from '..';
> 
> const client: AtlasClient = getMongodbAtlasApiClient({
>     privateKey: '',
>     publicKey: '',
>     projectId: '',
>     baseUrl: ''
> });
> 
> const newUser: CreateUserRequest = {
>     databaseName: 'db',
>     groupId: 'myGroup',
>     roles: [{
>         databaseName: 'db',
>         roleName: 'readWrite'
>     }],
>     password: 'usersPassword',
>     username: 'aUser'
> }
> 
> client.user.create(newUser).then(response => {
>     if ((response as AtlasError).errorCode) {
>         throw new Error('failed to create user');
>     }
>     const createdUser = (response as GetUserResponse);
>     console.log('created new user with access to: ', createdUser.databaseName)
> });
> ```
> 
> (My primary projects is still complaining but I think TypeScript is just confused in that case... some of this is still very confusing to me. haha)

I am happy to approve it and move forward. let me know

ccampanale commented:
I think that would be fine if you want to. Worst case scenario we go back and continue to make modifications. I feel like it's as correct as it will get for now least we get someone with a bit more TS experience to help out.

Also, I hadn't mentioned it before, but thanks for having created and supporting this package! Works well for our current use case. 👍

montumodi commented:
No worries. Glad it helped you and thanks again for rasing this PR.

montumodi commented:
published version `1.21.0`
